File: Python/check3.py

import sys
sys.stdin = open('../input.txt')

# MAXCITY = 100
# MAXCOUNT = 2000 + 1
INF = 999999
city_name = {}
trains = []
from_to = [[], []]  # fr_h, fr_t
ncity = 0
nconn = 0

station = ["Hakodate", "Tokyo"]
start = 8 * 60
end = 18 * 60
meetTime = 30

# 駅名 -> 駅番号


def city_id(name):
    global INF, ncity, nconn

    # keyが存在すればvalueを返す
    if name in city_name:
        return city_name[name]

    # 要素を追加
    city_name[name] = ncity
    ncity += 1
    return ncity - 1

# 列車追加


def parse_connection(buf):
    global INF, ncity, nconn

    # リスト内に辞書を追加
    trains.append({})
    trains[nconn]["from"] = city_id(buf[0])
    h, m = map(int, buf[1].split(":"))
    trains[nconn]["dpt"] = h * 60 + m
    trains[nconn]["to"] = city_id(buf[2])
    h, m = map(int, buf[3].split(":"))
    trains[nconn]["arv"] = h * 60 + m
    trains[nconn]["fare"] = int(buf[4])
    nconn += 1

# 列車配列tvの中で，dpttime以前に駅stに到着する列車を探す
# p: 探索を開始する列車到着事象の番号


def change(tr, p, st, dptTime):
    global INF, ncity, nconn

    while p >= 0:
        if (tr[p]["to"] == st) and (tr[p]["arv"] <= dptTime):
            break
        p -= 1
    return p

# 列車配列tvを参照して，駅orgを起点として，
# fr_h または fr_t を二次元配列vに作成する。


def make_tbl(tbl, org, tr):
    global INF, ncity, nconn

    # tblの初期化
    for i in range(0, ncity):
        tbl.append([])
        for _ in range(0, nconn + 1):
            tbl[i].append(INF)
    tbl[org][0] = 0

    # 列車の数だけ更新
    for ti in range(0, nconn):
        # 前の状態を引き継ぐ
        for i in range(0, ncity):
            tbl[i][ti + 1] = tbl[i][ti]

        # 列車配列tvの中で，tv[ti].dpt以前に駅tv[ti].fromに到着する列車を探す
        # つまり，列車tv[ti]に乗り継げる列車を探す。
        # ti-1: 探索を開始する列車到着事象の番号
        a = change(tr, ti - 1, tr[ti]["from"], tr[ti]["dpt"])

        # 安い方を引き継ぐ
        tbl[tr[ti]["to"]][ti + 1] = \
            min(tbl[tr[ti]["to"]][ti], tr[ti]
                ["fare"] + tbl[tr[ti]["from"]][a + 1])


def print_tbl(num):
    global INF, ncity, nconn, station

    print("From " if num < 2 else "To ", end="")
    print(station[num % 2], ":\ni\ttime|", end="")

    for i in range(0, ncity):
        print(i, "\t", end="")
    print("\n----------------+------------------------")
    for i in range(0, nconn + 1):
        print(i - 1, "\t", end="")
        if i:
            print(trains[i - 1]["arv"], end="")
        print("\t|", end="")
        for j in range(0, ncity):
            print(from_to[num][j][i], "\t", end="")
        print()

    print()


if __name__ == '__main__':
    while True:
        # 変数群のリセット
        ncity = 0
        nconn = 0
        trains.clear()
        city_name.clear()

        # 列車数読み込み
        count = int(input())
        print(count)

        # 終了条件
        if count == 0:
            break

        # 列車読み込み
        for i in range(0, count):
            buf = input().split()
            parse_connection(buf)

#         # 列車情報ソート前
#         print("Before:")
#         for i in range(0,nconn):
#             for j in trains[i]:
#                 print(trains[i][j]," ",end = "")
#             print()

        # "arv"をキーにしてソート
        trains.sort(key=lambda x: x["arv"])

#         # 列車情報ソート後
#         print("After:")
#         for i in range(0,nconn):
#             for j in trains[i]:
#                 print(trains[i][j]," ",end = "")
#             print()
#         print()

        # 表作成, 出力
        for i in range(0, len(from_to)):
            from_to[i].clear()
            make_tbl(from_to[i], city_id(station[i]), trains)
            print_tbl(i)

        print()

File: C/check3.c

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define FOR(i,a,b) for(int i=a;i<b;i++)
#define REP(i,n)  FOR(i,0,n)
#define MIN(a,b) (((a)<(b)) ? (a):(b))

#define MAXCITY 100
#define MAXCOUNT 2000 + 1
#define INFINITE 999999
char city_name[MAXCITY][18]; //駅名の表
int ncity; //駅数
int nconn; //列車数

int from_hakodate[MAXCITY][MAXCOUNT]; //fr_h
int from_tokyo[MAXCITY][MAXCOUNT];	 //fr_t

// 列車配列
struct train {
	int from, to;
	int dpt, arv;
	int fare;
} trains[MAXCOUNT];

//駅名 -> 駅番号
int city_id(char *name) {
	FOR(i,0,ncity) {
		if(strcmp(name, &(city_name[i][0])) == 0) return i;
	}
	strcpy(&(city_name[ncity][0]), name);
	return ncity++;
}

// 列車追加
void parse_connection(char *buf) {
	char from[18], to[18];
	int dpt[2], arv[2], fare;

	sscanf(buf,"%s %d:%d %s %d:%d %d",
			from, &(dpt[0]), &(dpt[1]), to, &(arv[0]), &(arv[1]), &fare);

	trains[nconn].from = city_id(from);
	trains[nconn].to = city_id(to);
	trains[nconn].dpt = dpt[0]*60 + dpt[1];
	trains[nconn].arv = arv[0]*60 + arv[1];
	trains[nconn].fare = fare;
	nconn++;
}

// qsort用関数（テキストの書き方は良くない）
int cmp_arv(const void *t1, const void *t2) {
    return ((struct train*)t1)->arv - ((struct train*)t2)->arv;
}

//到着時刻の早い順に並べ替え
void prepare_data(void) {
	qsort(trains, nconn, sizeof(struct train), cmp_arv);
}

//列車配列tvの中で，dpttime以前に駅stに到着する列車を探す
//p: 探索を開始する列車到着事象の番号
int change(struct train tv[], int p, int st, int dpttime) {
	while(p >= 0) {
		if((tv[p].to == st) && (tv[p].arv <= dpttime)) break;
		p--;
	}
	return p;
}

//列車配列tvを参照して，駅orgを起点として，
//fr_h または fr_t を二次元配列vに作成する。
void make_table(int v[MAXCITY][MAXCOUNT], int org, struct train tv[]) {
	int a;

	// 配列の初期化
	FOR(i,0,ncity) v[i][0] = INFINITE;
	v[org][0] = 0;

	FOR(ti,0,nconn) {
		// とりあえず今までの最安値を格納
		FOR(i,0,ncity) v[i][ti+1] = v[i][ti];

		//列車配列tvの中で，tv[ti].dpt以前に駅tv[ti].fromに到着する列車を探す
		//つまり，列車tv[ti]に乗り継げる列車を探す。
		//ti-1: 探索を開始する列車到着事象の番号
		a = change(tv, ti-1, tv[ti].from, tv[ti].dpt);

		// 今までの最安値と新しいルートでの金額を比較 & 安い方を採用
		v[tv[ti].to][ti+1] = MIN(v[tv[ti].to][ti], tv[ti].fare+v[tv[ti].from][a+1]);
	}
}

int main() {
	int count;
	char buf[64];

	count = 0;
	while(fgets(buf, sizeof(buf), stdin) != NULL) {
		// パラメータ初期化
		nconn = 0;
		ncity = 0;

		// 列車読み込み
		sscanf(buf, "%d", &count);
		printf("%d\n", count);
		if(!count) break;
		REP(i,count) {
			if(fgets(buf, sizeof(buf), stdin) == NULL) break;
			parse_connection(buf);
		}

		// 列車出力
		printf("Before:\n");
		REP(i,count)
			printf("%d %d %d %d %d\n",
					trains[i].from,trains[i].dpt,trains[i].to,trains[i].arv,trains[i].fare);
		// ソート
		prepare_data();

		// 列車出力
		printf("After:\n");
		REP(i,count)
			printf("%d %d %d %d %d\n",
				trains[i].from,trains[i].dpt,trains[i].to,trains[i].arv,trains[i].fare);

		// fr_hの作成
		make_table(from_hakodate, city_id("Hakodate"), trains);
		printf("From Hakodate:\ni\ttime |");
		REP(i,ncity) printf("%d	",i);
		printf("\n----------------+------------------------\n");
		REP(i,nconn+1) {
			printf("%d ",i-1);
			if(i) printf("%d\t",trains[i-1].arv); // 先頭行以外は到着時間出力
			printf("|");
			REP(j,ncity) printf("%d ",from_hakodate[j][i]);
			printf("\n");
		}
		printf("\n");

		// fr_tの作成
		make_table(from_tokyo, city_id("Tokyo"), trains);
		printf("From Tokyo:\ni\ttime |");
		REP(i,ncity) printf("%d\t",i);
		printf("\n----------------+------------------------\n");
		REP(i,nconn+1) {
			printf("%d ",i-1);
			if(i) printf("%d ",trains[i-1].arv);
			printf("|");
			REP(j,ncity) printf("%d ",from_tokyo[j][i]);
			printf("\n");
		}
		printf("\n\n");
	}
	return 0;
}






